### PR TITLE
[Geneva.Logs] Support prefix-based table name mapping

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

This PR adds support for prefix-based resolution (similar to [logging configuration](https://learn.microsoft.com/aspnet/core/fundamentals/logging/#configure-logging)) to the table mapping configuration options.

## Usage

Given this config...

```csharp
        var options = new GenevaExporterOptions
        {
            TableNameMappings = new Dictionary<string, string>()
            {
                ["*"] = "DefaultLogs",
                ["MyCompany"] = "InternalLogs",
                ["MyCompany.Product1"] = "InternalProduct1Logs",
                ["MyCompany.Product2"] = "InternalProduct2Logs",
                ["MyCompany.Product2.Security"] = "InternalSecurityLogs",
                ["MyPartner"] = "*",
            },
        };
```

* `ILogger<ThirdParty.Thing>`: This would go to "DefaultLogs"
* `ILogger<MyCompany.ProductX.Thing>`: This would go to "InternalLogs"
* `ILogger<MyCompany.Product1.Thing>`: This would go to "InternalProduct1Logs"
* `ILogger<MyCompany.Product2.Thing>`: This would go to "InternalProduct2Logs"
* `ILogger<MyCompany.Product2.Security.Alert>`: This would go to "InternalSecurityLogs"
* `ILogger<MyPartner.Product.Thing>`: This is marked as pass-through ("*") so it will be sanitized as "MyPartnerProductThing" table name

## Implementation details

The implementation is close to what ILogger does for its filter configuration. The idea is you don't want to have to apply the rules each time you process a log message because the perf would be terrible. You want to have the [result of rule selection](https://github.com/dotnet/runtime/blob/97f51dcca8e3e7ce85cbe52182db922f5afb90b0/src/libraries/Microsoft.Extensions.Logging/src/LoggerRuleSelector.cs#L10) computed once and available each time a log message is written. What the logging framework does is store the result on the `ILogger` instance created/injected from the factory. We don't have that luxury here. What I did was introduce a [helper class which maintains a cache](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/796/files#diff-0f16ceb5356860833a44dc4c125381d9560652433f15b48a504d6ad90f217209R39) of the final category -> table name. It is backed by a dictionary so you pay for a hash on the lookup which is what we were paying before (if mappings were used at all).

Some details about the cache...

* Lookups are lock-free. If a new mapping is needed we do a copy-on-write. So warmup is expensive but once you are warm it is cheap.

* There is no constraint on the number of keys (category names) which will be cached. ILogger itself maintains an unconstrained cache of loggers by category name so it seemed like that would OOM well before this.

* If the mapping is pass-through we only cache up to 1024 of the sanitized table names. Once the max is reached we just write them out to the buffer as they are logged (this was the existing behavior always).

## Benchmarks

Implementation + optimizations actually speed things up for everyone.

Before:
|                    Method | IncludeFormattedMessage |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------------------- |------------------------ |---------:|---------:|---------:|-------:|----------:|
| LoggerWithMessageTemplate |                   False | 546.6 ns | 10.43 ns |  9.76 ns | 0.0162 |     208 B |
| LoggerWithDirectLoggerAPI |                   False | 509.0 ns |  9.98 ns | 11.09 ns | 0.0267 |     344 B |
| LoggerWithSourceGenerator |                   False | 529.5 ns |  5.30 ns |  4.43 ns | 0.0134 |     168 B |
|        SerializeLogRecord |                   False | 359.4 ns |  5.70 ns |  5.34 ns | 0.0081 |     104 B |
|                    Export |                   False | 390.7 ns |  5.24 ns |  4.90 ns | 0.0081 |     104 B |
| LoggerWithMessageTemplate |                    True | 588.1 ns | 10.01 ns |  9.36 ns | 0.0162 |     208 B |
| LoggerWithDirectLoggerAPI |                    True | 512.0 ns |  7.44 ns |  6.96 ns | 0.0267 |     344 B |
| LoggerWithSourceGenerator |                    True | 537.0 ns |  8.15 ns |  6.80 ns | 0.0134 |     168 B |
|        SerializeLogRecord |                    True | 361.9 ns |  5.39 ns |  5.05 ns | 0.0081 |     104 B |
|                    Export |                    True | 391.7 ns |  4.14 ns |  3.87 ns | 0.0081 |     104 B |

After:
|                    Method | IncludeFormattedMessage |     Mean |    Error |  StdDev |   Gen0 | Allocated |
|-------------------------- |------------------------ |---------:|---------:|--------:|-------:|----------:|
| LoggerWithMessageTemplate |                   False | 530.6 ns |  3.30 ns | 2.92 ns | 0.0162 |     208 B |
| LoggerWithDirectLoggerAPI |                   False | 496.9 ns |  8.87 ns | 8.30 ns | 0.0267 |     344 B |
| LoggerWithSourceGenerator |                   False | 516.1 ns |  8.12 ns | 7.60 ns | 0.0134 |     168 B |
|        SerializeLogRecord |                   False | 343.4 ns |  6.64 ns | 6.52 ns | 0.0081 |     104 B |
|                    Export |                   False | 383.6 ns |  5.77 ns | 5.39 ns | 0.0081 |     104 B |
| LoggerWithMessageTemplate |                    True | 566.3 ns | 10.16 ns | 9.50 ns | 0.0162 |     208 B |
| LoggerWithDirectLoggerAPI |                    True | 504.9 ns |  9.36 ns | 8.29 ns | 0.0267 |     344 B |
| LoggerWithSourceGenerator |                    True | 524.2 ns |  5.07 ns | 4.75 ns | 0.0134 |     168 B |
|        SerializeLogRecord |                    True | 349.7 ns |  4.56 ns | 4.26 ns | 0.0081 |     104 B |
|                    Export |                    True | 380.9 ns |  3.93 ns | 3.67 ns | 0.0081 |     104 B |

## TODOs

* [X] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [X] Unit tests
* [X] README updates

### CODE REVIEW

File: src/OpenTelemetry.Exporter.Geneva/GenevaExporterOptions.cs [hunk 1]
```diff
@@ -47,14 +47,32 @@ public class GenevaExporterOptions
...
             {
+                if (string.IsNullOrWhiteSpace(entry.Key))
+                {
+                    throw new ArgumentException("A table name mapping key was null, empty, or consisted only of white-space characters.", nameof(this.TableNameMappings));
+                }
+
                 if (string.IsNullOrWhiteSpace(entry.Value))
                 {
-                    throw new ArgumentException($"A string-typed value provided for {nameof(this.TableNameMappings)} must not be null, empty, or consist only of white-space characters.");
+                    throw new ArgumentException($"The table name mapping value provided for key '{entry.Key}' was null, empty, or consisted only of white-space characters.", nameof(this.TableNameMappings));
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Geneva/GenevaExporterOptions.cs:57]: also checking heading/trailing spaces?

Comment (reyang) [reply src/OpenTelemetry.Exporter.Geneva/GenevaExporterOptions.cs:57]: never mind, I saw the IsValidTableName

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+        {
+            return tableName;
+        }
+
+        return this.ResolveTableMappingForCategoryNameRare(categoryName);
+    }
+
+    private byte[] ResolveTableMappingForCategoryNameRare(string categoryName)
+    {
+        lock (this.m_lockObject)
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:220]: Would a read-write-lock fit better here?

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+                }
+                else
+                {
+                    // Note: When the table name could not be sanitized we cache
+                    // the empty array NOT s_passthroughTableName.
+                    mappedTableName = Array.Empty<byte>();
+                }
+            }
+
+            // Note: This is using copy-on-write pattern to keep the happy
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:281]: 👍

File: src/OpenTelemetry.Exporter.Geneva/README.md [hunk 4]
```diff
@@ -104,14 +104,74 @@ sent through this exporter.
...
+  mode see below.
+
+* To change the table name for a specific log
+  [category](https://docs.microsoft.com/dotnet/core/extensions/logging#log-category)
+  add an entry with the key set to the full "category" of the log messages or a
+  prefix that will match the starting portion of the log message "category". Set
+  the value of the key to either the desired custom table name or `*` to enable
+  "pass-through" mapping. For details on "pass-through" mode see below.
+
+  For example, given the configuration...
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Geneva/README.md:134]: @ThomsonTan please review this part and see if the agent can do the same in theory (in case folks use the OTLP path and want to achieve the same).

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 3 above)
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:281]: Could we use ` ConcurrentDictionary`? The reads are lock-free for a `ConcurrentDictionary`.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+        return buffer;
+    }
+
+    // This method would map the logger category to a table name which only contains alphanumeric values with the following additions:
+    // Any character that is not allowed will be removed.
+    // If the resulting string is longer than 50 characters, only the first 50 characters will be taken.
+    // If the first character in the resulting string is a lower-case alphabet, it will be converted to the corresponding upper-case.
+    // If the resulting string still does not comply with Rule, the category name will not be serialized.
+    private static int SanitizedCategoryName(Span<byte> buffer, string categoryName)
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:158]: nit: Why remove the prefix `Serialize` from the method name?

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+
+            // If user configured table name mappings run resolution logic.
+            if (this.m_tableMappings != null
+                && !this.m_tableMappings.TryGetValue(categoryName, out mappedTableName))
+            {
+                // Find best match if an exact match was not found.
+
+                string currentKey = null;
+
+                foreach (var mapping in this.m_tableMappings)
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:241]: This can be optimized. We can avoid checking every key of the `TableNameMappings`, if we keep a collection of the keys sorted in descending order and use that to decide if we need to keep checking for next keys.

Let's consider the example provided in the description. Let's say the category name was `MyCompany.Product2.Component` and we have the keys sorted in descending order:
```text
"MyPartner"
"MyCompany.Product2.Security"
"MyCompany.Product2"
"MyCompany.Product1"
"MyCompany"
"*"
```
Once we find a match we can break out of the loop as that would be the longest match.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 5 above)
Comment (utpilla) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:158]: Never mind! Just noticed that this method does not always write to the "buffer". We could probably call the method `WriteSanitizedCategoryName` 😀

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+
+    private sealed class TableNameCacheDictionary : Dictionary<string, byte[]>, ITableNameCacheDictionary
+    {
+        public TableNameCacheDictionary()
+            : base(0, StringComparer.OrdinalIgnoreCase)
+        {
+        }
+
+        public TableNameCacheDictionary(TableNameCacheDictionary sourceCache)
+            : base(sourceCache, StringComparer.OrdinalIgnoreCase)
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:306]: `TableNameMappings` uses `StringComparer.Ordinal`.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+                        mappedTableName = mapping.Value;
+                    }
+                }
+            }
+
+            mappedTableName ??= !this.m_shouldPassThruTableMappings
+                ? this.m_defaultTableName
+                : s_passthroughTableName;
+
+            if (mappedTableName.Length <= 0 && tableNameCache.CachedSanitizedTableNameCount < MaxCachedSanitizedTableNames)
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:260]: Why does this limit of 1024 matter? If I understand correctly, regardless of the `CachedSanitizedTableNameCount` value, we would always add a new category as a key to the `m_tableNameCache`. So why not just add the actual sanitized category name as the value instead of `s_passthroughTableName` when the 1024 limit is reached.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 5 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:158]: I'll rename to `WriteSanitizedCategoryNameToSpan` 🤣

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 7 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:306]: Thanks, didn't notice that. Looked into it a bit. I was going to change to `OrdinalIgnoreCase` because it seems low-risk and nice for users but looking at the code `Ordinal` is used for all the lookup dictionaries. I'm going to set as `Ordinal` for now and look at that as a follow-up. I did refactor a bit so it would be easier to change in the future.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 8 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:262]: Short answer: Today we don't cache any of the sanitized names so this was me being conservative. 

Long answer:

> If I understand correctly, regardless of the CachedSanitizedTableNameCount value, we would always add a new category as a key to the m_tableNameCache

Correct

> So why not just add the actual sanitized category name as the value instead of s_passthroughTableName when the 1024 limit is reached.

Just to make it less greedy. Each category added after 1024 will grow by a pointer to s_passthroughTableName instead of a pointer to some freshly allocated buffer (up to 52 bytes IIRC).

Comment (utpilla) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:262]: > Just to make it less greedy. Each category added after 1024 will grow by a pointer to s_passthroughTableName instead of a pointer to some freshly allocated buffer (up to 52 bytes IIRC).

Would 52 bytes be considered high? For every new set of 1024 keys, it's only 52 KB max.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+                continue;
+            }
+
+            return false;
+        }
+
+        return true;
+    }
+
+    public int ResolveAndSerializeTableNameForCategoryName(byte[] destination, int offset, string categoryName, out ReadOnlySpan<byte> tableName)
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:122]: Use `MethodImplOptions.AggressiveInlining`?

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:223]: Could we use read-write-lock or concurrent dictionary (as @utpilla suggested below) here? Yes. BUT... the performance is not identical. My goal here is to not force a perf hit on users happy without the prefix feature, if possible. My thinking with this design was based on the idea that if we took a limit to infinity, there are no writes to the dictionary at all 🤣 The writes only exist as a blip during warmup. So my goal was to keep the reads as fast as possible once everything is warm. With concurrent dictionary and `ReaderWriterLockSlim` there seems to be overhead. `Hashtable` is really close and avoids the need for copy-on-write. I'm considering making a change to use `Hashtable`.

|               Method |      Mean |     Error |    StdDev |
|--------------------- |----------:|----------:|----------:|
|           Dictionary |  9.505 ns | 0.2339 ns | 0.2188 ns |
| ConcurrentDictionary | 15.491 ns | 0.3507 ns | 0.3281 ns |
|           ReaderLock | 26.550 ns | 0.5548 ns | 0.7595 ns |
|            Hashtable | 10.890 ns | 0.2053 ns | 0.1920 ns |

<details>
<summary>Benchmark code</summary>

```csharp
using System;
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using BenchmarkDotNet.Attributes;

namespace OpenTelemetry.Exporter.Geneva.Benchmark;

public class DictionaryBenchmarks
{
    private readonly ReaderWriterLockSlim readerWriterLock = new();

    private readonly Dictionary<string, byte[]> values = new(StringComparer.Ordinal)
    {
        ["Category1"] = new byte[1],
        ["Category2"] = new byte[1],
        ["Category3"] = new byte[1],
    };

    private readonly ConcurrentDictionary<string, byte[]> cvalues = new(StringComparer.Ordinal)
    {
        ["Category1"] = new byte[1],
        ["Category2"] = new byte[1],
        ["Category3"] = new byte[1],
    };

    private readonly Hashtable hvalues = new(StringComparer.Ordinal)
    {
        ["Category1"] = new byte[1],
        ["Category2"] = new byte[1],
        ["Category3"] = new byte[1],
    };

    [Benchmark]
    public byte[] Dictionary()
    {
        if (this.values.TryGetValue("Category1", out var value))
        {
            return value;
        }

        lock (this.values)
        {
            return LookupCategoryNameRare("Category1");
        }
    }

    [Benchmark]
    public byte[] ConcurrentDictionary()
    {
        return this.cvalues.GetOrAdd("Category1", LookupCategoryNameRare);
    }

    [Benchmark]
    public byte[] ReaderLock()
    {
        this.readerWriterLock.EnterReadLock();
        try
        {
            if (this.values.TryGetValue("Category1", out var value))
            {
                return value;
            }
        }
        finally
        {
            this.readerWriterLock.ExitReadLock();
        }

        this.readerWriterLock.EnterWriteLock();
        try
        {
            return LookupCategoryNameRare("Category1");
        }
        finally
        {
            this.readerWriterLock.ExitWriteLock();
        }
    }

    [Benchmark]
    public byte[] Hashtable()
    {
        byte[] mapping = (byte[])this.hvalues["Category1"];
        if (mapping != null)
        {
            return mapping;
        }

        lock (this.hvalues)
        {
            return LookupCategoryNameRare("Category1");
        }
    }

    private static byte[] LookupCategoryNameRare(string categoryName)
    {
        throw new NotSupportedException("All the benchmarks should hit the cache.");
    }
}
```
</details>

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:284]: @utpilla I replied to this above.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 8 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:263]: Honestly, didn't think too much about it. Just nice to prevent unconstrained growth where we can. The check is only processed the first time we see a category name value and then we cache the result. So it is pretty inexpensive (a few nanos) to have the cap. You want me to remove it?

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 9 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:125]: Updated

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:286]: Mentioned above possibly using `Hashtable`. [Hashtable has a unique quirk in that it supports the many-readers-single-writer pattern](https://learn.microsoft.com/dotnet/api/system.collections.hashtable#thread-safety). That would allow removing copy-on-write, which is great for the warm-up phase:

|      Method |       Mean |    Error |   StdDev |      Gen0 |     Gen1 |   Allocated |
|------------ |-----------:|---------:|---------:|----------:|---------:|------------:|
| CopyOnWrite | 3,659.8 us | 63.39 us | 56.19 us | 1328.1250 | 183.5938 | 16326.84 KB |
|   Hashtable |   103.6 us |  1.95 us |  1.82 us |   13.4277 |   6.5918 |    165.3 KB |

<details>
<summary>Benchmark code</summary>

```csharp
using System.Collections;
using System.Collections.Generic;
using BenchmarkDotNet.Attributes;

namespace OpenTelemetry.Exporter.Geneva.Benchmark;

[MemoryDiagnoser]
public class CopyOnWriteBenchmarks
{
    private readonly object lockObject = new();

    [Benchmark]
    public Dictionary<string, byte[]> CopyOnWrite()
    {
        Dictionary<string, byte[]> d = new();

        for (int i = 0; i < 1024; i++)
        {
            string k = $"Category{i}";

            if (!d.TryGetValue(k, out var value))
            {
                lock (this.lockObject)
                {
                    if (!d.TryGetValue(k, out value))
                    {
                        d[k] = new byte[1];

                        var old = d;

                        d = new(old);
                    }
                }
            }
        }

        return d;
    }

    [Benchmark]
    public Hashtable Hashtable()
    {
        Hashtable ht = new();

        for (int i = 0; i < 1024; i++)
        {
            string k = $"Category{i}";

            var value = (string)ht[k];
            if (value == null)
            {
                lock (this.lockObject)
                {
                    value = (string)ht[k];
                    if (value == null)
                    {
                        ht[k] = new byte[1];
                    }
                }
            }
        }

        return ht;
    }
}
```
</details>

Chatted with @utpilla we decided to not use `Hashtable` for now because the docs also say it isn't recommended to be used.

It is an implementation detail so we can always change it in the future.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 8 above)
Comment (utpilla) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:265]: More than the `if` check I was thinking we could save on another round of parsing of Category name when we have more than 1024 category names.

It will come down to the cost of copying a small `byte[]` to the buffer vs executing WriteSanitizedCategoryName (parsing + serialize). I was wondering if we anyway have to parse the string for dictionary lookup why not just get hold of the `byte[]` instead of again parsing and serializing the sanitized Category name.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 2 above)
Comment (utpilla) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs:225]: > My goal here is to not force a perf hit on users happy without the prefix feature, if possible.

That's a good point.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+                {
+                    // Note: When the table name could not be sanitized we cache
+                    // the empty array NOT s_passthroughTableName.
+                    mappedTableName = Array.Empty<byte>();
+                }
+            }
+
+            // Note: This is using copy-on-write pattern to keep the happy
+            // path lockless once everything has spun up.
+            this.m_tableNameCache = new(tableNameCache)
```
[the remark this answers is not in the document]
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs]: Updated

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 8 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs]: Hmm. I think there is some confusion about what is going on. So we always cache _something_. For a given category name the cache will be one of these things...

1) A defined mapping (pointer to something in m_tableMappings).
2) The default table name (pointer to m_defaultTableName).
3) A sanitized category name (allocation).
4) Array.Empty<byte> if category name could not be sanitized (pointer to static).
5) The pass-through indicator if the cache was full (pointer to s_passthroughTableName).

The logic is...

* Lookup the mapping using category name from the cache.
* If the mapping is the pass-through indicator, write out the sanitized table name directly to the buffer.

For pass-through today the logic is...

* Check for a mapping using category name.
* If nothing was found and pass-through is enabled, write out the sanitized table name directly to the buffer.

So basically when the cache fills up we don't do double-work, we just fallback to the behavior we have today.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 6 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs]: I've been working on this but it is getting kind of big with benchmarks, etc. Thinking I'll merge this PR and then open perf improvements as its own thing.

File: src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs (hunk 8 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs]: Update... I synced up with @utpilla offline. The concern here is that in computing hash code for CategoryName in order to look into the cache "parses" the string. I have some ideas to improve this, but I don't thing we need to do anything here. The perf should be better for most, slightly worse for some.

* For users using at least 1 mapping, there is already a hash code computation + dictionary lookup in their path today. What these users will see: There should be see a speed-up for <= 1024 categories. Same speed as today for > 1024 categories.

* For users NOT using any mappings today, there is now a new hash code computation + dictionary lookup in their path. What these users will see: When using default table name, there should be a speed-up because we now copy a pre-built array using span helpers instead of calling into `MessagePackSerializer.SerializeAsciiString` each time. When using pass-through: Speed-up for <= 1024 categories. A slight slow-down for > 1024 categories.
